client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,178 @@
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		onSuccessRedirectDestination: PropTypes.string,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+			isRedirecting: false,
+			newForwardAdded: false,
+		};
+	}
+
+	hasValidEmailForwards() {
+		return ! this.state.emailForwards?.some( ( forward ) => ! forward.isValid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination ) => {
+		withAnalytics(
+			composeAnalytics(
+				this.props.onConfirmEmailForwarding(),
+				this.props.addEmailForward( domainName, mailbox, destination )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { selectedSiteSlug, selectedDomainName } = this.props;
+
+		event.preventDefault();
+
+		if ( this.props.isAddingEmailForward ) {
+			return;
+		}
+
+		this.state.emailForwards?.map( ( forward ) => {
+			const { mailbox, destination } = forward;
+
+			this.addNewEmailForwardWithAnalytics(
+				selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+		this.state.newForwardAdded = true;
+	};
+
+	onForwardAdd = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
+		} );
+	};
+
+	renderActionsButtons() {
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				<Button
+					primary
+					onClick={ this.addNewEmailForwardsClick }
+					disabled={ ! this.hasValidEmailForwards() }
+				>
+					{ translate( 'Add' ) }
+				</Button>
+			</div>
+		);
+	}
+
+	onRemoveEmailForward = ( index ) => {
+		const [ emailForwards ] = this.state;
+		emailForwards.splice( index, 1 );
+		this.setState( { emailForwards } );
+	};
+
+	onUpdateEmailForward = ( index, name, value ) => {
+		// eslint-disable-next-line prefer-const
+		let emailForwards = this.state.emailForwards;
+		emailForwards[ index ][ name ] = value;
+
+		const validEmailForward = validateAllFields( emailForwards[ index ] );
+		emailForwards[ index ].isValid =
+			validEmailForward.mailbox.length === 0 && validEmailForward.destination.length === 0;
+
+		this.setState( { emailForwards } );
+	};
+
+	componentDidUpdate() {
+		const {
+			emailForwardSuccess,
+			isRequestingDomains,
+			onSuccessRedirectDestination,
+			siteId,
+		} = this.props;
+
+		const { isRedirecting, newForwardAdded } = this.state;
+
+		if (
+			emailForwardSuccess &&
+			onSuccessRedirectDestination &&
+			! isRedirecting &&
+			newForwardAdded
+		) {
+			if ( ! isRequestingDomains && ! isRedirecting ) {
```

Comment:
I am on the fence about whether this code should always trigger the domains fetch. I have two concerns:
* the receiving component should generally be ensuring that it has the correct data
* do we always need to trigger a fetch? Maybe it’s only needed for the first one. (I am not sure about this, to be honest.)

Reply:
I am not sure what to do here. I had in mind that just after adding new forwards, our store would already have updated data. As this wasn't the case, we need to update our data as soon as we have added a new forward, so this is the case. Also I see this fetch a bit "anti pattern" but I can't think a better way to do so.

We could fetch always in the emails/home or emails/management but I think it's not nice to fetch if we haven't change anything in the client.